hibernate-core/src/main/java/org/hibernate/Query.java

```diff
@@ -64,6 +65,106 @@
 	 */
 	String getQueryString();
 
+	/**
+	 * "QueryOptions" is a better name, I think, than "RowSelection" -> 6.0
+	 *
+	 * @todo 6.0 rename RowSelection to QueryOptions
+	 *
+	 * @return Return the encapsulation of this query's options, which includes access to
+	 * firstRow, maxRows, timeout and fetchSize.   Important because this gives access to
+	 * those values in their Integer form rather than the primitive form (int) required by JPA.
+	 */
+	RowSelection getQueryOptions();
+
+	/**
+	 * The position of the first query result to be retrieved, previously set by
+	 * {@link #setFirstResult(int)} or {@link #setHibernateFirstResult(int)}.
+	 * <p/>
+	 * If the value was not initialized by {@link #setFirstResult(int)} or
+	 * {@link #setHibernateFirstResult(int)}, then {@code null} is returned, resulting
+	 * in pagination starting from position 0.
+	 * <p/>
+	 * If {@link #setHibernateFirstResult(int)} was called with a negative value, then 0
+	 * is returned.
+	 *
+	 * @return the position of the first query result, or {@code null} if uninitialized.
+	 *
+	 * @see #setFirstResult(int)
+	 * @see #setHibernateFirstResult(int)
+	 *
+	 * @deprecated {@link #setFirstResult(int)} should be used instead.
+	 */
+	@Deprecated
+	default Integer getHibernateFirstResult() {
+		return getQueryOptions().getFirstRow();
+	}
+
+	/**
+	 * Set the position of the first query result to be retrieved. A negative value will
+	 * result in pagination starting from position 0.
+	 *
+	 * @param firstRow - the position of the first query result
+	 * @return {@code this}, for method chaining
+	 *
+	 * @deprecated {@link #setFirstResult(int)} should be used instead.
+	 */
+	@Deprecated
+	default Query setHibernateFirstResult(int firstRow) {
+		if ( firstRow < 0 ) {
+			getQueryOptions().setFirstRow( 0 );
+		}
+		else {
+			getQueryOptions().setFirstRow( firstRow );
+		}
+		return this;
+	}
+
+	/**
+	 * The maximum number of query results to be retrieved, previously set by
+	 * {@link #setMaxResults(int)} or {@link #setHibernateMaxResults(int)}.
+	 * <p/>
+	 * If the value was not initialized by {@link #setMaxResults(int)} or
+	 * {@link #setHibernateMaxResults(int)}, then {@code null} is returned
+	 * <p/>
+	 * If {@link #setHibernateMaxResults(int)} was called with a value less than
+	 * or equal to 0, the value is considered to be uninitialized, and {@code null}
+	 * is returned, resulting in no limit on the number of results.
+	 *
+	 * @return the maximum number of query results, or {@code null} if uninitialized.
+	 *
+	 * @see #setFirstResult(int)
+	 * @see #setHibernateFirstResult(int)
+	 *
+	 * @deprecated {@link #setFirstResult(int)} should be used instead.
+	 */
+	@Deprecated
+	default Integer getHibernateMaxResults() {
+		return getQueryOptions().getMaxRows();
+	}
+
+	/**
+	 * Set the maximum number of query results to be retrieved. A value less than
+	 * or equal to 0 is considered uninitialized, resulting in no limit on the number
```

Comment:
This whole thing that "0" means "all" was clearly a bug in previous versions. I think it's ok to try to maintain the semantics, but should we really commit this promise into documentation? Not sure if sets the wrong expectations.

Reply:
@Sanne, why do you think it was a bug?

Reply:
In any case, this method is deprecated, so we are only committing to it until it is removed.

Reply:
@Sanne, IMO, for existing applications, the difference between all results and no results is too much of a behavioral difference.

An alternative that I would find acceptable (not sure about anyone else) is for Hibernate to log a warning if ```setMaxResults( 0 )``` is called on an object that does *not* extend ```java.persistence.Query```. WDYT?